python/lsst/ip/diffim/imageMapReduce.py

```diff
@@ -622,146 +622,121 @@ def _generateGrid(self, exposure, forceEvenSized=False, **kwargs):
-        if gridStepX > bbox.getWidth():
-            gridStepX = bbox.getWidth()
-        gridStepY = rescaleValue(gridStepY)
-        if gridStepY > bbox.getHeight():
-            gridStepY = bbox.getHeight()
-        borderSizeX = rescaleValue(borderSizeX)
-        borderSizeY = rescaleValue(borderSizeY)
-
-        nGridX = bbox.getWidth() // gridStepX
-        nGridY = bbox.getHeight() // gridStepY
-
-        if adjustGridOption == 'spacing':
-            nGridX = bbox.getWidth() / gridStepX
-            # Readjust gridStepX so that it fits perfectly in the image.
-            gridStepX = float(bbox.getWidth() - gridSizeX) / float(nGridX)
-            if gridStepX < 1:
-                raise ValueError('X grid spacing is too small (or negative): %f' % gridStepX)
-
-        if adjustGridOption == 'spacing':
-            nGridY = bbox.getWidth() / gridStepY
-            # Readjust gridStepY so that it fits perfectly in the image.
-            gridStepY = float(bbox.getHeight() - gridSizeY) / float(nGridY)
-            if gridStepY < 1:
-                raise ValueError('Y grid spacing is too small (or negative): %f' % gridStepY)
-
-        if adjustGridOption == 'size':
-            gridSizeX = gridStepX
-            gridSizeY = gridStepY
-
-        print('Grid parameters:', gridSizeX, gridSizeY, gridStepX, gridStepY, borderSizeX, borderSizeY)
+                cellCentroids = [(x, y) for x in np.arange(cellSizeX//2, bbox.getWidth() + cellSizeX//2,
+                                                           gridStepX)
+                                        for y in np.arange(cellSizeY//2, bbox.getHeight() + cellSizeY//2,
+                                                           gridStepY)]
+
+            cellCentroidsX, cellCentroidsY = zip(*cellCentroids)
 
         # first "main" box at 0,0
-        bbox0 = afwGeom.Box2I(afwGeom.Point2I(bbox.getBegin()), afwGeom.Extent2I(gridSizeX, gridSizeY))
+        bbox0 = afwGeom.Box2I(afwGeom.Point2I(bbox.getBegin()), afwGeom.Extent2I(cellSizeX, cellSizeY))
         # first expanded box
         bbox1 = afwGeom.Box2I(bbox0)
         bbox1.grow(afwGeom.Extent2I(borderSizeX, borderSizeY))
 
         self.boxes0 = []  # "main" boxes; store in task so can be extracted if needed
         self.boxes1 = []  # "expanded" boxes
 
-        # use given centroids as centers for bounding boxes
-        if self.config.gridCentroidsX is not None and len(self.config.gridCentroidsX) > 0:
-            for i, centroidX in enumerate(self.config.gridCentroidsX):
-                centroidY = self.config.gridCentroidsY[i]
+        def _makeBoxEvenSized(bb):
+            """Force a bounding-box to have dimensions that are modulo 2."""
+
+            if bb.getWidth() % 2 == 1:  # grow to the right
+                bb.include(afwGeom.Point2I(bb.getMaxX()+1, bb.getMaxY()))  # Expand by 1 pixel!
+                bb.clip(bbox)
+                if bb.getWidth() % 2 == 1:  # clipped at right -- so grow to the left
```

Comment:
Hmm, it's only possible if the cell size is > the image size. This shouldn't happen but I suppose if it does I can shrink the `bb` instead of growing it. 

Ugh, `Box2D`s are not very easily manipulatable. I'm just going to raise an exception if it is not even-sized after these two attempts, let's see how often it actually happens.